django/thunderstore/api/cyberstorm/views/package_preview.py

```diff
@@ -0,0 +1,288 @@
+import math
+from itertools import chain
+from typing import OrderedDict
+
+from django.contrib.postgres.aggregates import ArrayAgg
+from django.core.paginator import EmptyPage, Page
+from django.db.models import (
+    Case,
+    CharField,
+    Count,
+    Lookup,
+    Q,
+    QuerySet,
+    Sum,
+    Value,
+    When,
+)
+from django.http import HttpRequest, HttpResponse
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import status
+from rest_framework.exceptions import ParseError
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from thunderstore.api.cyberstorm.serializers.package import (
+    CommunityPackageListCyberstormSerializer,
+    PackageCategoryCyberstormSerializer,
+    PackageSearchQueryParameterCyberstormSerializer,
+)
+from thunderstore.cache.enums import CacheBustCondition
+from thunderstore.cache.pagination import CachedPaginator
+from thunderstore.community.models import PackageListingSection
+from thunderstore.community.models.package_category import PackageCategory
+from thunderstore.community.models.package_listing import PackageListing
+
+
+class LowerContainedBy(Lookup):
+    lookup_name = "icontained_by"
+
+    def as_sql(self, compiler, connection):
+        lhs, lhs_params = self.process_lhs(compiler, connection)
+        rhs, rhs_params = self.process_rhs(compiler, connection)
+        params = lhs_params + rhs_params
+        return f"LOWER({rhs}) LIKE '%%' || LOWER({lhs}) || '%%'", params
+
+
+CharField.register_lookup(LowerContainedBy)
+
+
+class PackagePreviewAPIView(APIView):
+    permission_classes = []
+
+    @swagger_auto_schema(
+        responses={200: CommunityPackageListCyberstormSerializer()},
+        operation_id="api.packages",
+    )
+    def get(
+        self,
+        request: HttpRequest,
+    ) -> HttpResponse:
+
+        qp = PackageSearchQueryParameterCyberstormSerializer(data=request.query_params)
+        if not qp.is_valid():
+            return Response(qp.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        params: OrderedDict = qp.validated_data
+        listings_qs = self.get_queryset(params)
+        listings_qs = self.order_queryset(params["ordering"], listings_qs)
+        categories = PackageCategory.objects.filter(
+            pk__in=listings_qs.aggregate(arr=ArrayAgg("categories", distinct=True))[
+                "arr"
+            ]
+        )
```

Comment:
Just include all the categories that are in use in the community, no need to be fancy about this.